chore: update gh workflows with zizmor suggested fixes

mthorning opened the pull request:
This pull request updates multiple GitHub Actions workflows to improve security, stability, and functionality. The most notable changes include pinning action versions to specific commit SHAs, adding validation for branch names, and updating dependencies and configurations.

### Security Improvements:
* Added `persist-credentials: false` to `actions/checkout` steps in `.github/workflows/build-engine-docker-image-and-publish-to-dockerhub.yml` and `.github/workflows/e2e-tests.yml` to reduce the risk of unauthorized access to credentials. [[1]](diffhunk://#diff-f87667d48e22abfbbccf3695a8acc386143e3542286f432e6e3d8330960c76f9R29) [[2]](diffhunk://#diff-194218c48b9a0cdd03974145733804c2d992ca818529fe2fa69a501d8b5b1cc3R48-R59)
* Validated branch names against a safe pattern in `.github/workflows/linting-and-tests.yml` to prevent potential misuse of branch names in subsequent steps.

### Stability Enhancements:
* Pinned all third-party GitHub Actions to specific commit SHAs across various workflows to ensure consistent and predictable behavior. Examples include `grafana/shared-workflows/actions/build-push-to-dockerhub`, `catchpoint/workflow-telemetry-action`, and `actions-ecosystem/action-remove-labels`. [[1]](diffhunk://#diff-f87667d48e22abfbbccf3695a8acc386143e3542286f432e6e3d8330960c76f9L37-R38) [[2]](diffhunk://#diff-194218c48b9a0cdd03974145733804c2d992ca818529fe2fa69a501d8b5b1cc3R48-R59) [[3]](diffhunk://#diff-f93a3de9563193d65121683e6383741ac4b6aa18bdb51ba82b80497e700561cdL15-R15)
* Updated Helm-related actions in `.github/workflows/linting-and-tests.yml` and `.github/workflows/on-release-published.yml` to specific SHAs for better reliability. [[1]](diffhunk://#diff-a70d3d29c45894eeef2036c533385dbc424f9479590aaea01e62c06dc67079a1L147-R170) [[2]](diffhunk://#diff-e95a5d3f03a1351728732657b6b150cfbbd9a9724b387226b1f99f079b1954b0L91-R91)

### Functional Updates:
* Enhanced `.github/workflows/linting-and-tests.yml` by using validated branch references in Git commands to avoid errors caused by unsafe branch names.
* Updated `snyk/actions/setup` in `.github/workflows/snyk-security-scan.yml` to a specific SHA for improved compatibility and security.

These changes collectively enhance the security, reliability, and maintainability of the workflows.

jdbaldry reviewed (commented):
The publish technical documentation and update-make-docs workflows will be continuously updated by Renovate if you're using it until I get https://github.com/renovatebot/renovate/pull/35431 merged because it won't understand the comment.

mthorning commented:
> The publish technical documentation and update-make-docs workflows will be continuously updated by Renovate if you're using it until I get [renovatebot/renovate#35431](https://github.com/renovatebot/renovate/pull/35431) merged because it won't understand the comment.

Sorry @jdbaldry, can you explain? I've updated the workflow to use the commit hash directly instead of the tag, but is there something else happening that I'm not aware of?

jdbaldry commented:
> > The publish technical documentation and update-make-docs workflows will be continuously updated by Renovate if you're using it until I get [renovatebot/renovate#35431](https://github.com/renovatebot/renovate/pull/35431) merged because it won't understand the comment.
> 
> Sorry @jdbaldry, can you explain? I've updated the workflow to use the commit hash directly instead of the tag, but is there something else happening that I'm not aware of?

sorry for the lack of context. Your change indeed pins to a commit hash as you intended. Thanks to the use of mutable tags for GitHub Action versioning, your comment will drift out of date when I release newer v1 versions of the Action so you may want to consider marking the specific minor and patch version you've pinned to.

If you're intending to track the tag using Renovate to automatically open pull requests to update the commit hash, that won't currently work because my tag naming convention doesn't match the regular expression Renovate uses (mine has a slash, they expect a dash). If you are using Renovate, because it can't understand the tag comment, then it will assume that you're tracking the `main` branch, and you'll get a PR every time there's commits to the `main` branch of Writers' Toolkit.

brojd commented on .github/workflows/on-release-published.yml:
```diff
@@ -120,7 +120,7 @@ jobs:
     steps:
       - name: Get Vault secrets
-        uses: grafana/shared-workflows/actions/get-vault-secrets@main
+        uses: grafana/shared-workflows/actions/get-vault-secrets@75804962c1ba608148988c1e2dc35fbb0ee21746
```
Can we pls add a comment with the version that the commit hash is related to?

mthorning replied:
These are the current latest commit in main though that will obviously no longer be the case when the next commit is made. Do you think I should pin them to the latest release version instead?

mthorning replied:
I see now, I didn't realise the comments were used by another tool, I'm updating all to match `vX.X.X`

mthorning commented:
> > > The publish technical documentation and update-make-docs workflows will be continuously updated by Renovate if you're using it until I get [renovatebot/renovate#35431](https://github.com/renovatebot/renovate/pull/35431) merged because it won't understand the comment.
> > 
> > 
> > Sorry @jdbaldry, can you explain? I've updated the workflow to use the commit hash directly instead of the tag, but is there something else happening that I'm not aware of?
> 
> sorry for the lack of context. Your change indeed pins to a commit hash as you intended. Thanks to the use of mutable tags for GitHub Action versioning, your comment will drift out of date when I release newer v1 versions of the Action so you may want to consider marking the specific minor and patch version you've pinned to.
> 
> If you're intending to track the tag using Renovate to automatically open pull requests to update the commit hash, that won't currently work because my tag naming convention doesn't match the regular expression Renovate uses (mine has a slash, they expect a dash). If you are using Renovate, because it can't understand the tag comment, then it will assume that you're tracking the `main` branch, and you'll get a PR every time there's commits to the `main` branch of Writers' Toolkit.

Ah, I see. Sorry, I wasn't aware of this functionality! So, if I need to update the commit hash to a tag with the full semantic versioning and then change the comment so it's like `#update-make-docs/v1.X.X`?

mthorning commented:
@jdbaldry @brojd I have updated the version tags, please let me know if this is correct now. I also reverted a change which had been made that would have subtly changed the way the files were diffed. Thank you 🙏
